Objects/funcobject.c

```diff
@@ -859,24 +881,41 @@ func_set_annotate(PyObject *self, PyObject *value, void *Py_UNUSED(ignored))
     }
 }
 
+/*[clinic input]
+@critical_section
+@getter
+function.__annotations__
+
+Dict of annotations in a function object.
+[clinic start generated code]*/
+
 static PyObject *
-func_get_annotations(PyObject *self, void *Py_UNUSED(ignored))
-{
-    PyFunctionObject *op = _PyFunction_CAST(self);
-    if (op->func_annotations == NULL &&
-        (op->func_annotate == NULL || !PyCallable_Check(op->func_annotate))) {
-        op->func_annotations = PyDict_New();
-        if (op->func_annotations == NULL)
+function___annotations___get_impl(PyFunctionObject *self)
+/*[clinic end generated code: output=a4cf4c884c934cbb input=92643d7186c1ad0c]*/
+{
+    PyObject *d = NULL;
+    Py_BEGIN_CRITICAL_SECTION(self);
```

Comment:
If the function is annotated with `@critical_section`, we shouldn't use `Py_BEGIN_CRITICAL_SECTION(self)` calls within it.